File: plantmaintenance/plantmaintenance/page/inventory_approval/inventory_approval.js
frappe.pages['inventory-approval'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Inventory Approval',
        single_column: true 
    });

    let $task_container = $('<div>').appendTo(page.body);

    let $table = $('<table class="table table-bordered">').appendTo($task_container);
    $table.append(`
        <thead>
            <tr>
                <th><input type="checkbox" id="select-all"></th> <!-- Master checkbox -->
                <th>No</th> 
                <th>Task Name</th>
                <th>Material Code</th>
                <th>Material Name</th>
                <th>Available Quantity</th>
                <th>Required Quantity</th>
                <th>Shortage</th>
                <th>Status</th>
                <th>Approval Date</th>
                <th>Issued Date</th>
            </tr>
        </thead>
        <tbody></tbody>
    `);

    let $tbody = $table.find('tbody');
    let index = 1; 

    frappe.call({
        method: 'plantmaintenance.plantmaintenance.page.inventory_approval.inventory_approval.get_task_details',
        callback: function(response) {
            if (response.message) {
                response.message.forEach(function(task_detail) {
                    if (task_detail.materials.length > 0) {
                        task_detail.materials.forEach(function(material) {
                            $tbody.append(`
                                <tr>
                                    <td><input type="checkbox" class="material-checkbox" data-task="${task_detail.task_name}" data-material="${material.material_code}"></td>
                                    <td>${index++}</td>
                                    <td>${task_detail.task_name}</td>
                                    <td>${material.material_code}</td>
                                    <td>${material.material_name}</td>
                                    <td>${material.available_quantity}</td>
                                    <td>${material.required_quantity}</td>
                                    <td>${material.shortage}</td>
                                    <td>${material.status}</td>
                                    <td>${frappe.datetime.str_to_user(material.approval_date)}</td>
                                    <td>${frappe.datetime.str_to_user(material.issued_date)}</td>
                                </tr>
                            `);
                        });
                    }
                });
            }
        }
    });

    let $approve_button = $('<button class="btn btn-primary">Approve</button>').appendTo($task_container);

    $approve_button.on('click', function() {
        let selectedMaterials = [];
        $task_container.find('.material-checkbox:checked').each(function() {
            let row = $(this).closest('tr');
            selectedMaterials.push({
                task_name: row.find('td:eq(2)').text(),
                material_code: row.find('td:eq(3)').text(),
                available_quantity: row.find('td:eq(5)').text(),
                required_quantity: row.find('td:eq(6)').text()
            });
        });

        if (selectedMaterials.length > 0) {
            frappe.call({
                method: 'plantmaintenance.plantmaintenance.page.inventory_approval.inventory_approval.approve_materials',
                args: {
                    materials: selectedMaterials
                },
                callback: function(response) {
                    if (response.message) {
                        frappe.msgprint(response.message);
                    }
                }
            });
        } else {
            frappe.msgprint('No materials selected.');
        }
    });

    $task_container.on('change', '#select-all', function() {
        let isChecked = $(this).prop('checked');
        $task_container.find('.material-checkbox').prop('checked', isChecked);
    });
};

File: plantmaintenance/plantmaintenance/page/inventory_approval/inventory_approval.py
# import frappe
# import json
# from datetime import datetime

# @frappe.whitelist()
# def get_task_details():
#     user = frappe.session.user
#     tasks = frappe.get_all('Task Detail', filters={'approver': user}, fields=['name'])

#     task_details = []
    
#     for task in tasks:
#         task_materials = frappe.get_all('Material Issue', filters={
#             'parent': task.name, 
#             'status': 'Pending Approval'
#         }, fields=[
#             'material_code', 'material_name', 'available_quantity', 'required_quantity', 'shortage', 'status', 'approval_date'
#         ])
#         if task_materials:
#             task_details.append({
#                 'task_name': task.name,
#                 'materials': task_materials
#             })

#     return task_details


# @frappe.whitelist()
# def approve_materials(materials):
#     if not isinstance(materials, list):
#         materials = json.loads(materials)

#     today = datetime.today().strftime('%Y-%m-%d')
    
#     for item in materials:
#         task_name = item.get('task_name')
#         material_code = item.get('material_code')

#         task_doc = frappe.get_all(
#             "Task Detail",
#             filters={"name": task_name},
#             fields=["name"]
#         )
        
#         if task_doc:
#             task_doc = frappe.get_doc("Task Detail", task_name)
            
#             for material in task_doc.material_issued:
#                 if (material.material_code == material_code and
#                     material.status == "Pending Approval"):
#                     material.status = "Material Issued"
#                     material.issued_date = today
                    
#                     task_doc.append("material_returned", {
#                         "material_code": material.material_code,
#                         "material_name": material.material_name,
#                         "issue_quantity": material.required_quantity,
#                         "approval_date": material.approval_date,
#                         "issued_date": today
#                     })
                    
#                     task_doc.save()
#                     break  

#     return "Selected materials successfully approved."


import frappe
import json
from datetime import datetime

@frappe.whitelist()
def get_task_details():
    user = frappe.session.user
    tasks = frappe.get_all('Task Detail', filters={'approver': user}, fields=['name'])

    task_details = []
    
    for task in tasks:
        task_materials = frappe.get_all('Material Issue', filters={
            'parent': task.name, 
            'status': 'Pending Approval'
        }, fields=[
            'material_code', 'material_name', 'available_quantity', 'required_quantity', 'shortage', 'status', 'approval_date'
        ])
        if task_materials:
            task_details.append({
                'task_name': task.name,
                'materials': task_materials
            })

    return task_details

@frappe.whitelist()
def approve_materials(materials):
    if not isinstance(materials, list):
        materials = json.loads(materials)

    today = datetime.today().strftime('%Y-%m-%d')
    
    for item in materials:
        task_name = item.get('task_name')
        material_code = item.get('material_code')

        task_doc = frappe.get_all(
            "Task Detail",
            filters={"name": task_name},
            fields=["name"]
        )
        
        if task_doc:
            task_doc = frappe.get_doc("Task Detail", task_name)
            
            for material in task_doc.material_issued:
                if (material.material_code == material_code and
                    material.status == "Pending Approval"):
                    material.status = "Material Issued"
                    material.issued_date = today
                    
                    task_doc.append("material_returned", {
                        "material_code": material.material_code,
                        "material_name": material.material_name,
                        "issue_quantity": material.required_quantity,
                        "approval_date": material.approval_date,
                        "issued_date": today
                    })
                    
                    task_doc.save()
                    break  

    return "Selected materials successfully approved."

@frappe.whitelist()
def reject_materials(materials):
    if not isinstance(materials, list):
        materials = json.loads(materials)
    
    for item in materials:
        task_name = item.get('task_name')
        material_code = item.get('material_code')

        task_doc = frappe.get_all(
            "Task Detail",
            filters={"name": task_name},
            fields=["name"]
        )
        
        if task_doc:
            task_doc = frappe.get_doc("Task Detail", task_name)
            
            for material in task_doc.material_issued:
                if (material.material_code == material_code and
                    material.status == "Pending Approval"):
                    material.status = "Material Rejected"
                    
                    task_doc.save()
                    break  

    return "Selected materials successfully rejected."
